File: Blatt2/ex2.py
import numpy as np
import matplotlib.pyplot as plt

h = 0.0001
t_max = 5
phi_start = 1
l = 1
m = 1
g = 9.81

phi_starts = [0.01, np.pi / 4, np.pi / 2, 3 * np.pi / 4]

class RungeKutta_4:
    def __init__(self, f, h):
        self.f = f
        self.h = h

    def __k1(self, y, t):
        return self.f(y, t) * self.h

    def __k2(self, y, t):
        return self.f(y + 0.5 * self.__k1(y, t), t + 0.5 * self.h) * self.h

    def __k3(self, y, t):
        return self.f(y + 0.5 * self.__k2(y, t), t + 0.5 * self.h) * self.h

    def __k4(self, y, t):
        return self.f(y + self.__k3(y, t), t + self.h) * self.h

    def advance_y(self, y, t):
        return y + 1 / 6 * (self.__k1(y, t) + 2 * self.__k2(y, t) + 2 * self.__k3(y, t) + self.__k4(y, t))
    

def get_wave_period(time, result):
    downwards = True
    start_time = 0
    cross_zero = 0
    for index, point in enumerate(result):
        if point < 0 and downwards:
            cross_zero += 1
            downwards = False
        if point > 0 and not downwards:
            downwards = True
        if cross_zero == 1 and start_time == 0:
            start_time = time[index]
        elif cross_zero == 2:
            return time[index] - start_time
        

def force(phi):
        return - m * g * np.sin(phi)

def f(y, t):
    return np.array([y[1], force(y[0])])

def analytical(t, phi):
    return phi * np.cos(np.sqrt(g / l) * t)

def plot(phi):
    y = np.array([phi, 0])

    R = RungeKutta_4(f, h)

    result = []
    result_analytical = []
    time = []

    for i in range(int(t_max / h)):
        t = i * h
        time.append(t)
        result.append(y[0])
        a = analytical(t, phi)
        result_analytical.append(a)
        y = R.advance_y(y, t)

    print('Wave period numeric ({}): {}s'.format(phi, get_wave_period(time, result)))
    print('Wave period analytical ({}): {}s'.format(phi, get_wave_period(time, result_analytical)))

    p1 = plt.plot(time, result)
    p2 = plt.plot(time, result_analytical)
    plt.title('Phi_0 = {}'.format(phi))
    plt.xlabel('time/s')
    plt.ylabel('phi')
    plt.legend((p1[0], p2[0]), ('Numeric result', 'Analytical result'))
    plt.show()

def main():
    for i in phi_starts:
        plot(i)

main()

File: Blatt2/ex2.cpp
#include <iostream>
#include <cmath>
#include <vector>

using namespace std;

const double m = 1;
const double l = 1;
const double g = 9.81;
const double h = 0.01;

const double t_max = 10;

double force(double phi)
{
    return m * g * std::sin(phi);
}

vector<double> get_f(double y[2], double t)
{
    vector<double> result = {y[1], force(y[0])};
    return result;
}

vector<double> euler(double phi_start)
{
    double phi = phi_start;
    double y[] = {phi, 0};
    vector<double> results(t_max / h);

    for (double t = 0; t < t_max; t += h)
    {
        results[t / h] = y[0];
        vector<double> f = get_f(y, t);
        y[0] += f[0] * h;
        y[1] += f[1] * h;
    }

    return results;
}

vector<double> get_k2(double y[2], double t)
{
    vector<double> k1 = get_f(y, t);
    y[0] += 0.5 * k1[0] * h;
    y[1] += 0.5 * k1[1] * h;
    vector<double> result = get_f(y, t + 0.5 * h);
    result[0] += result[0] * h;
    result[1] += result[1] * h;
    return result;
}

vector<double> get_k3(double y[2], double t)
{
    vector<double> k2 = get_k2(y, t);
    y[0] += 0.5 * k2[0] * h;
    y[1] += 0.5 * k2[1] * h;
    vector<double> result = get_f(y, t + 0.5 * h);
    return result;
}

vector<double> get_k4(double y[2], double t)
{
    vector<double> k3 = get_k3(y, t);
    y[0] += k3[0] * h;
    y[1] += k3[1] * h;
    vector<double> result = get_f(y, t + h);
    return result;
}

vector<double> runge_4(double phi_start)
{
    double phi = phi_start;
    double y[] = {phi, 0};
    vector<double> results(t_max / h);

    for (double t = 0; t < t_max; t += h)
    {
        results[t / h] = y[0];
        vector<double> f = get_f(y, t);
        y[0] += f[0] * h;
        y[1] += f[1] * h;
    }

    return results;
}

int main()
{
    double phi_start = 1;

    vector<double> euler_results = euler(phi_start);
}

